platforms/core-runtime/client-services/src/main/java/org/gradle/launcher/daemon/toolchain/DaemonJavaToolchainProvisioningService.java

```diff
@@ -74,14 +74,25 @@ public File tryInstall(JavaToolchainSpec spec) {
             try {
                 File downloadFolder = cacheDirProvider.getDownloadLocation();
                 ExternalResource resource = downloader.getResourceFor(uri);
-                File archiveFile = new File(downloadFolder, getFileName(uri, resource));
+                File archiveFile = new File(downloadFolder, getUniqueFileName(uri, resource, spec));
                 final FileLock fileLock = cacheDirProvider.acquireWriteLock(archiveFile, "Downloading toolchain");
+                boolean canSkipDownload = archiveFile.exists();
                 try {
-                    if (!archiveFile.exists()) {
+                    if (!canSkipDownload) {
                         downloader.download(uri, archiveFile, resource);
                     }
                     progressLogger.progress("Unpacking toolchain archive " + archiveFile.getName());
-                    File installedToolchainFile = cacheDirProvider.provisionFromArchive(spec, archiveFile, uri);
+                    File installedToolchainFile;
+                    try {
+                        installedToolchainFile = cacheDirProvider.provisionFromArchive(spec, archiveFile, uri);
+                    } catch (Exception e) {
+                        if (canSkipDownload) { // re-download and retry in case the archive is corrupted
+                            downloader.download(uri, archiveFile, resource);
+                            installedToolchainFile = cacheDirProvider.provisionFromArchive(spec, archiveFile, uri);
+                        } else {
+                            throw e;
+                        }
+                    }
```

Comment:
❗ The error-handling here seems too broad.  On any exception (out of space, permissions issues, etc.) we immediately try re-downloading?  Can this be narrowed to the cases where re-downloading has a chance of solving the problem and won't just add extra wait before an unresolvable issue is reported to the user?

Reply:
💭 It would also be nice to rename `canSkipDownload` to `archiveFound` or something, as it's odd to see `if-canSkipDownload-then-download`, which looks like a typo at first glance.

Reply:
> It would also be nice to rename canSkipDownload to archiveFound or something,

Thanks, I had the same thought and struggled to find a good name.

> The error-handling here seems too broad.

That's intentional. I think it's hard to categorize every possible type of issue here. Sure, it's a performance issue, but it only happens on failure. If it were there, the original issue about archives having the same name wouldn't be a problem for the user. And there are scenarios where the issues you listed as an example may be caused by a bad archive. I think UX is much more valuable here than machine time.

Reply:
That makes sense, and I suppose the extra wait on a build that is doomed to fail anyway isn't such a big concern.  It would be nice if `provisionFromArchive` method did some work to categorize failures and throw different types of exceptions that would allow different responses, but I agree it's not a showstopper here, maybe an idea for a future improvement.

What about the possibility of the failure recovery download/provision failing and hiding the original exception from the first attempt at provisioning?  If provisioning fails in a _different_ way the original failure isn't reported.  This does seem like something to adjust now, but putting the recovery code inside its own try-catch, or maybe a separate method.